### ISSUE TITLE
[FirebaseCore] java.lang.NullPointerException crash on Goolge Play console

### ISSUE THREAD

**Issue by MAC-Joyco** [myflashlab/Firebase-ANE#131]
Hi!
I'm getting the following crash report on the Google Play console from about the 1% of my users:

java.lang.NullPointerException: 
  at com.myflashlab.firebase.core.AirCommand.onConnectionFailed (AirCommand.java:407)
  at com.google.android.gms.common.internal.zzac.zzk (Unknown Source)
  at com.google.android.gms.internal.zzbcp.zzc (Unknown Source)
  at com.google.android.gms.internal.zzbcd.zze (Unknown Source)
  at com.google.android.gms.internal.zzbcd.zza (Unknown Source)
  at com.google.android.gms.internal.zzbch.zzpV (Unknown Source)
  at com.google.android.gms.internal.zzbcy.zzc (Unknown Source)
  at com.google.android.gms.internal.zzbcz.handleMessage (Unknown Source)
  at android.os.Handler.dispatchMessage (Handler.java:102)
  at android.os.Looper.loop (Looper.java:148)
  at android.app.ActivityThread.main (ActivityThread.java:7409)
  at java.lang.reflect.Method.invoke (Native Method)
  at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run (ZygoteInit.java:1230) 
  at com.android.internal.os.ZygoteInit.main (ZygoteInit.java:1120)

This is the issue distribution by devices and OS:

Grand Prime Plus (grandpplte) 		8.2%
Galaxy J7(2016) (j7xelte) 		7.8%
Galaxy J5 Prime (on5xelte) 		6.7%
Galaxy J5(2016) (j5xnlte) 		6.1%
Others 		71.3%

Android 6.0 		57.8%
Android 5.1 		23.4%
Android 4.4 		9.6%
Android 5.0 		5.0%
Others 		4.2%

But I can not reproduce the problem in my tests.
I'm using Firebase Core + Authentication + Database and the ANEs I've added for this are:

* androidSupport.ane V24.2.1
* overrideAir.ane v4.0.0
* firebase_common.ane V11.0.2
* firebase_iid.ane V11.0.2
* googlePlayServices_base.ane V11.0.2
* googlePlayServices_basement.ane V11.0.2
* googlePlayServices_tasks.ane V11.0.2
* googlePlayServices_appinvite.ane V11.0.2
* Firebase Core v5.0.0
* firebase_auth.ane V11.0.2
* firebase_database.ane V11.0.2
* firebase_databaseConnection.ane V11.0.2

And this is an excerpt from my manifest (I have replaced sensitive info by xxxxx):

```xml
	<manifest android:installLocation="auto">
		<uses-sdk android:minSdkVersion="15" android:targetSdkVersion="19" />

		<uses-permission android:name="android.permission.INTERNET"/>
		<uses-permission android:name="android.permission.WRITE_EXTERNAL_STORAGE"/>
		<uses-permission android:name="android.permission.WAKE_LOCK"/>
		<uses-permission android:name="android.permission.ACCESS_NETWORK_STATE"/>
		<uses-permission android:name="android.permission.ACCESS_WIFI_STATE"/>
		<uses-permission android:name="android.permission.RECEIVE_BOOT_COMPLETED"/>
		<uses-permission android:name="com.android.vending.BILLING" />
			    			    
		<!-- 	Required by firebase_iid.ane 
			Change "air.com.doitflash.firebaseCore" to your own app package name
		-->
		<uses-permission android:name="com.google.android.c2dm.permission.RECEIVE" />
		<permission android:name="air.com.xxxxxxxx.xxxxx.permission.C2D_MESSAGE" android:protectionLevel="signature" />
		<uses-permission android:name="air.com.xxxxxxxxxx.xxxxx.permission.C2D_MESSAGE" />
				
		<application android:enabled="true" android:largeHeap="true" android:isGame="true">
			<activity android:excludeFromRecents="false">
				<intent-filter>
					<action android:name="android.intent.action.MAIN"/>
					<category android:name="android.intent.category.LAUNCHER"/>
				</intent-filter>
						
				<intent-filter> 
					<action android:name="android.intent.action.VIEW"/> 
					<category android:name="android.intent.category.BROWSABLE"/> 
					<category android:name="android.intent.category.DEFAULT"/> 
					<data android:scheme="fbxxxxxxxxxxxxxxx"/>
					<data android:scheme="air.com.xxxxx.xxxx" /> 
				</intent-filter> 
                        						
			</activity>

			<!-- 	Required by the firebase_common.ane 
				Change "air.com.doitflash.firebaseCore" to your own app package name
			-->
			<provider
				android:name="com.google.firebase.provider.FirebaseInitProvider"
				android:authorities="air.com.xxxxxx.xxxxx.firebaseinitprovider"
				android:exported="false"
				android:initOrder="100" />
		
			<!--
				Required by googlePlayServices_basement.ane
        		-->        				
        		<meta-data android:name="com.google.android.gms.games.APP_ID" android:value="\ xxxxxxxxxxxxxx" />
        		<meta-data android:name="com.google.android.gms.version" android:value="@integer/google_play_services_version" />
		
			<!-- 	Required by firebase_iid.ane 
				Change "air.com.doitflash.firebaseCore" to your own app package name
			-->
			<receiver
				android:name="com.google.firebase.iid.FirebaseInstanceIdReceiver"
				android:exported="true"
				android:permission="com.google.android.c2dm.permission.SEND" >
				<intent-filter>
					<action android:name="com.google.android.c2dm.intent.RECEIVE" />
			                <!-- <action android:name="com.google.android.c2dm.intent.REGISTRATION" /> -->
			                <category android:name="air.com.xxxxxxx.xxxx" />
				</intent-filter>
			</receiver>
			<receiver android:name="com.google.firebase.iid.FirebaseInstanceIdInternalReceiver" android:exported="false" />
			<service android:name="com.google.firebase.iid.FirebaseInstanceIdService" android:exported="true">
				<intent-filter android:priority="-500">
					<action android:name="com.google.firebase.INSTANCE_ID_EVENT" />
				</intent-filter>
			</service>
		
			<!--	Required by googlePlayServices_base.ane
			-->
			<activity android:name="com.google.android.gms.common.api.GoogleApiActivity"
				android:theme="@android:style/Theme.Translucent.NoTitleBar"
				android:exported="false"/>
					
		</application>
	</manifest>
```

Thanks in advance!

**Comment by myflashlab** [comment]
Your setting seems fine but the main question is why those devices cannot connect to google inner APIs and when they don't, why the crash happens!

I wonder if you can somehow identify those devices and get some more information from them using the debug tool in the [Override ANE](https://github.com/myflashlab/common-dependencies-ANE/tree/master/overridAir)?

**Comment by MAC-Joyco** [comment]
Thanks for your quick answer!

To activate the debugger I would have to be able to reproduce the error on my devices, but at the moment I have not been able to.
On the other hand, I wonder what the impact of this problem is on users, since so far I have not received any complaint review. By the type of error and the place where it occurs, what are the possible consequences of this error? Does the application close after this error? Or does the execution continue causing a simple Firebase connection error event?

There are also other variants of the problem in the same line of code but with slightly different stack trace, and much lower incidence, for example:

java.lang.NullPointerException: 
  at com.myflashlab.firebase.core.AirCommand.onConnectionFailed (AirCommand.java:407)
  at com.google.android.gms.common.internal.zzac.zzk (Unknown Source)
  at com.google.android.gms.internal.zzbcp.zzc (Unknown Source)
  at com.google.android.gms.internal.zzbcd.zze (Unknown Source)
  at com.google.android.gms.internal.zzbcd.onConnectionSuspended (Unknown Source)
  at com.google.android.gms.internal.zzbcx.onConnectionSuspended (Unknown Source)
  at com.google.android.gms.internal.zzbbi.onConnectionSuspended (Unknown Source)
  at com.google.android.gms.common.internal.zzaa.onConnectionSuspended (Unknown Source)
  at com.google.android.gms.common.internal.zzh.handleMessage (Unknown Source)
  at android.os.Handler.dispatchMessage (Handler.java:102)
  at android.os.Looper.loop (Looper.java:135)
  at android.app.ActivityThread.main (ActivityThread.java:5376)
  at java.lang.reflect.Method.invoke (Native Method)
  at java.lang.reflect.Method.invoke (Method.java:372)
  at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run (ZygoteInit.java:908)
  at com.android.internal.os.ZygoteInit.main (ZygoteInit.java:703)

java.lang.NullPointerException: 
  at com.myflashlab.firebase.core.AirCommand.onConnectionFailed (AirCommand.java:407)
  at com.google.android.gms.common.internal.zzac.zzk (Unknown Source)
  at com.google.android.gms.internal.zzbcp.zzc (Unknown Source)
  at com.google.android.gms.internal.zzbcd.zze (Unknown Source)
  at com.google.android.gms.internal.zzbcd.zzpW (Unknown Source)
  at com.google.android.gms.internal.zzbcd.zza (Unknown Source)
  at com.google.android.gms.internal.zzbcx.zza (Unknown Source)
  at com.google.android.gms.internal.zzbbi.onConnectionFailed (Unknown Source)
  at com.google.android.gms.common.internal.zzab.onConnectionFailed (Unknown Source)
  at com.google.android.gms.common.internal.zzn.zzj (Unknown Source)
  at com.google.android.gms.common.internal.zze.zzs (Unknown Source)
  at com.google.android.gms.common.internal.zzi.zzrk (Unknown Source)
  at com.google.android.gms.common.internal.zzh.handleMessage (Unknown Source)
  at android.os.Handler.dispatchMessage (Handler.java:110)
  at android.os.Looper.loop (Looper.java:193)
  at android.app.ActivityThread.main (ActivityThread.java:5299)
  at java.lang.reflect.Method.invokeNative (Native Method)
  at java.lang.reflect.Method.invoke (Method.java:515)
  at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run (ZygoteInit.java:831) 
  at com.android.internal.os.ZygoteInit.main (ZygoteInit.java:647)
  at dalvik.system.NativeStart.main (Native Method)

java.lang.NullPointerException: 
  at com.myflashlab.firebase.core.AirCommand.onConnectionFailed (AirCommand.java:407)
  at com.google.android.gms.common.internal.zzac.zzk (Unknown Source)
  at com.google.android.gms.internal.zzbcp.zzc (Unknown Source)
  at com.google.android.gms.internal.zzbcd.zze (Unknown Source)
  at com.google.android.gms.internal.zzbcd.zzpW (Unknown Source)
  at com.google.android.gms.internal.zzbcd.zzk (Unknown Source)
  at com.google.android.gms.internal.zzbcf.zzf (Unknown Source)
  at com.google.android.gms.common.internal.zzo.zzj (Unknown Source)
  at com.google.android.gms.common.internal.zze.zzs (Unknown Source)
  at com.google.android.gms.common.internal.zzi.zzrk (Unknown Source)
  at com.google.android.gms.common.internal.zzh.handleMessage (Unknown Source)
  at android.os.Handler.dispatchMessage (Handler.java:102)
  at android.os.Looper.loop (Looper.java:148)
  at android.app.ActivityThread.main (ActivityThread.java:5523)
  at java.lang.reflect.Method.invoke (Native Method)
  at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run (ZygoteInit.java:799) 
  at com.android.internal.os.ZygoteInit.main (ZygoteInit.java:689)

Which suggests that there may be several causes of this problem, and while they are fixed, I think it would be a good idea to implement a more secure programming in that method. For example, if the error is causing the app to close and since the error is a java.lang.NullPointerException at com.myflashlab.firebase.core.AirCommand.onConnectionFailed (AirCommand.java:407), it should be easy to avoid the exception by checking that the state of the variables involved in that line of code are not null, and returning the corresponding connection error event if it is the case.

Regards!

**Comment by xangel82** [comment]
I have the same error at the line 603, please can you manage the exception by checking that the state of the variables involved in that line of code are not null.

Exception java.lang.NullPointerException: Attempt to invoke virtual method 'android.app.Dialog com.google.android.gms.common.GoogleApiAvailability.getErrorDialog(android.app.Activity, int, int, android.content.DialogInterface$OnCancelListener)' on a null object reference
com.myflashlab.firebase.core.AirCommand.onConnectionFailed (AirCommand.java:603)
com.google.android.gms.common.internal.zzae.zzk ()
com.google.android.gms.common.api.internal.zzbd.zzc ()
com.google.android.gms.common.api.internal.zzar.zze ()
com.google.android.gms.common.api.internal.zzar.zza ()
com.google.android.gms.common.api.internal.zzav.zzahq ()
com.google.android.gms.common.api.internal.zzbm.zzc ()
com.google.android.gms.common.api.internal.zzbn.handleMessage ()
android.os.Handler.dispatchMessage (Handler.java:102)
android.os.Looper.loop (Looper.java:148)
android.app.ActivityThread.main (ActivityThread.java:5551)
java.lang.reflect.Method.invoke (Method.java)

**Comment by CrazyFlasher** [comment]
same shit

**Comment by myflashlab** [comment]
Ok, although we have not been able to duplicate this problem, but it seems like we have a fix for it. Considering that this error is originated in the native dialog somehow, we have removed the dialog and will dispatch the content back to AIR with this listener ```GOOGLE_API_CONNECTION_FAILURE```.

We have released FirebaseCore V6.0.1 to ANELAB. If you are a club member, you can simply update the ANE through ANELAB and test if the problem is resolved. If not, you have to wait a little longer until we finish our next round of updates to the Firebase collection (which includes the Crashlytics ANE replacing the current Crash ANE).

Let us know how does your tests go.

**Comment by myflashlab** [comment]
**NOTICE:** If you need MyFlashLabs developers' professional help, you should be either a [club member](https://www.myflashlabs.com/product/myflashlabs-club-membership/) or have a [premium support package](https://www.myflashlabs.com/product/myflashlabs-support/).
